keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,733 @@
+  - type: Webhook
+    webhook:
+      name: in-cluster-authorizer
+      authorizedTTL: 5m
+      unauthorizedTTL: 30s
+      timeout: 3s
+      subjectAccessReviewVersion: v1
+      failurePolicy: NoOpinion
+      connectionInfo:
+        type: InClusterConfig
+```
+
+Validation will allow multiple authorizers of type "Webhook" to be added to the
+config, but one authorizer each for other types. The ordering of this chain will
+be decided by the order specified in the file.
+
+The keys `kubeConfigFile`, `authorizedTTL`, `unauthorizedTTL` and
+`subjectAccessReviewVersion` accept values corresponding to flags
+`--authorization-webhook-config-file`, `--authorization-webhook-cache-authorized-ttl`,
+`--authorization-webhook-cache-unauthorized-ttl` and `--authorization-webhook-version`
+respectively.
+
+Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding
+flag is not supplied. While configuring authorization modes using the file config,
+the version supported by a webhook has to be mentioned using a required field
+`subjectAccessReviewVersion`.
+
+The user can define a CEL expression to determine whether a request needs to dispatched
+to the authz webhook for which the expression has been defined. The user would have access
+to a `request` variable containing a `SubjectAccessReview` object in the version specified
+by `subjectAccessReviewVersion`.
+
+The code path for enabling the above will only be triggered if the feature flag will
+be enabled until the time the feature flag is removed and configuring authorizer
+through a file becomes GA.
+
+### Monitoring
+
+We will add the following 4 metrics:
+
+1. `apiserver_authorization_step_invocations_total`
+
+This will be incremented on round-trip of an authorizer. It will track total
+authorization decision invocations across the following labels.
+
+Labels {along with possible values}:
+- `mode` {<authorizer_name>}
+- `decision` {Allow, Deny, NoOpinion}
+
+**Note:** Some examples of <authorizer_name>: `RBAC`, `Node`, `ABAC`, `webhook{,_<name>}`.
+If there is only one webhook, there would be no `_<name>` suffix.
+
+2. `apiserver_authorization_step_webhook_invocations_total`
+
+This will be incremented on round-trip of an authorization webhook. It will track
+total invocation counts across the following labels.
+
+- `name`
+- `code` {2xx, 4xx, 5xx}
+- `decision` {Allow, Deny, NoOpinion}
+
+3. `apiserver_authorization_step_webhook_duration_total_seconds`
+
+This metric will track the total round trip time of the requests to the webhook.
+
+Labels {along with possible values}:
+- `name`
+- `code` {2xx, 4xx, 5xx}
+- `decision` {Allow, Deny, NoOpinion}
+
+4. `apiserver_authorization_step_webhook_error_total`
+
+This metric will be incremented when a webhook returns a 4xx or 5xx (erroneous) response.
+
+Labels {along with possible values}:
+
+- `name`
+- `code` {4xx, 5xx}
+- `decision` {Deny, NoOpinion}
+
+5. `apiserver_authorization_step_configuration_reload_failure_count`
```

Comment:
Shouldn't this be Gauge instead of count so we can see last timestamp for current state?

How about `apiserver_authorization_step_configuration_reload_last_timestamp_seconds`
```
The last time in seconds when an authorization reload was performed.
- `name`
- `status` {success, fail}
```

Reply:
I agree. A gauge is more appropriate.